adaptive/learner/average_mixin.py

```diff
@@ -0,0 +1,202 @@
+        if x in self.pending_points:
+            self.pending_points[x].discard(seed)
+            if not self.pending_points[x]:
+                # pending_points[x] is now empty so delete the set()
+                del self.pending_points[x]
+
+    def _add_to_data(self, point, value):
+        x, seed = self.unpack_point(point)
+        if x not in self._data:
+            self._data[x] = DataPoint()
+        self._data[x][seed] = value
+
+    def ask(self, n, tell_pending=True):
+        """Return n points that are expected to maximally reduce the loss."""
+        points, loss_improvements = self._ask_points_without_adding(n)
+        loss_improvements = self._normalize_new_points_loss_improvements(
+            points, loss_improvements)
+
+        p, l = self.loss_per_existing_point()
+        l = self._normalize_existing_points_loss_improvements(p, l)
+        points += p
+        loss_improvements += l
+
+        loss_improvements, points = zip(*sorted(
+            zip(loss_improvements, points), reverse=True))
+
+        points = list(points)[:n]
+        loss_improvements = list(loss_improvements)[:n]
+
+        if tell_pending:
+            for p in points:
+                self.tell_pending(p)
+
+        return points, loss_improvements
+
+    def n_values(self, point):
+        pending_points = self.pending_points.get(point, [])
+        return len(self._data[point]) + len(pending_points)
+
+    def _mean_values_per_neighbor(self, neighbors):
+        """The average number of neighbors of a 'point'."""
+        return {p: sum(self.n_values(n) for n in ns) / len(ns)
+            for p, ns in neighbors.items()}
+
+    def _normalize_new_points_loss_improvements(self, points, loss_improvements):
+        """If we are suggesting a new point, then its 'loss_improvement' should
+        be divided by the average number of values of its neigbors."""
+        if len(self._data) < 4:
+            return loss_improvements
+
+        only_points = [p for p, s in points]
+        neighbors = self._get_neighbor_mapping_new_points(only_points)
+        mean_values_per_neighbor = self._mean_values_per_neighbor(neighbors)
+
+        return [loss / mean_values_per_neighbor[p]
+            for (p, seed), loss in zip(points, loss_improvements)]
+
+    def _normalize_existing_points_loss_improvements(self, points, loss_improvements):
+        """If the neighbors of 'point' have twice as much values
+        on average, then that 'point' should have an infinite loss."""
+        if len(self._data) < 4:
+            return loss_improvements
+
+        neighbors = self._get_neighbor_mapping_existing_points()
+        mean_values_per_neighbor = self._mean_values_per_neighbor(neighbors)
+
+        def needs_more_data(p):
+            return mean_values_per_neighbor[p] > 1.5 * self.n_values(p)
+
+        return [inf if needs_more_data(p) else loss
+            for (p, seed), loss in zip(points, loss_improvements)]
+
+
+def add_average_mixin(cls):
```

Comment:
I'm not sure whether this is clear enough; feels like way too much magic. And from what I see, the usual inheritance should be good enough.

Reply:
> the usual inheritance should be good enough

That was not the case. I don't remember the exact details but there were issues with the method resolution order, there was no other option than doing it like this.

If you want I could find out why exactly.